server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1131,72 +1430,232 @@ public static boolean useIndexGenerations(Version repositoryMetaVersion) {
+
             @Override
             public ClusterState execute(ClusterState currentState) {
                 SnapshotDeletionsInProgress deletions = currentState.custom(SnapshotDeletionsInProgress.TYPE);
                 if (deletions != null) {
                     boolean changed = false;
                     if (deletions.hasDeletionsInProgress()) {
-                        assert deletions.getEntries().size() == 1 : "should have exactly one deletion in progress";
-                        SnapshotDeletionsInProgress.Entry entry = deletions.getEntries().get(0);
-                        deletions = deletions.withRemovedEntry(entry);
-                        changed = true;
+                        final SnapshotDeletionsInProgress updatedDeletions = deletions.withRemovedEntry(deleteEntry.uuid());
+                        changed = updatedDeletions != deletions;
+                        deletions = updatedDeletions;
+                        if (failure == null) {
+                            // The delete worked out so we remove the snapshot ids that it removed from the repository from queued up
+                            // delete jobs
+                            deletions = deletions.withRemovedSnapshotIds(deleteEntry.repository(), deleteEntry.getSnapshots());
+                        }
                     }
                     if (changed) {
-                        return ClusterState.builder(currentState).putCustom(SnapshotDeletionsInProgress.TYPE, deletions).build();
+                        // We removed a delete from the cluster state. If that delete caused any snapshots to not have their shard
+                        // snapshots assigned and we have to assign them now based on the new RepositoryData that resulted from the delete.
+                        final SnapshotsInProgress snapshotsInProgress = currentState.custom(SnapshotsInProgress.TYPE);
+                        // There is one special case to deal with here: If the RepositoryData passed to this method is null, then that
+                        // means that the delete operation failed to load the current RepositoryData, indicating an issue with the
+                        // repository so we fail all queued up operations.
+                        final boolean failAllQueuedOperations = repositoryData == null;
+                        if (failAllQueuedOperations) {
+                            deletions.getEntries().stream().filter(entry -> entry.repository().equals(deleteEntry.repository()))
+                                    .map(SnapshotDeletionsInProgress.Entry::uuid).forEach(deletionsToFail::add);
+                            // We failed to read repository data for this delete, it is not the job of SnapshotsService to
+                            // retry this kind of issue so we fail all the pending deletes
+                            deletions = deletions.withRemovedRepository(deleteEntry.repository());
```

Comment:
Should there be a return after this line?

I see usage later on of repositoryData where it's possibly null

Reply:
Also sounds like we have a testing gap

Reply:
I think we're safe here because we don't do any actual shard assignments that would use `RepositoryData` if it's `null`, we just break out here: https://github.com/elastic/elasticsearch/pull/56911/files#diff-a0853be4492c052f24917b5c1464003dR1529 and fail all the queued up snapshots and deletes. This is actually covered by e.g. test `org.elasticsearch.snapshots.ConcurrentSnapshotsIT#testQueuedDeletesWithFailures` that has both a queued up snapshot and delete fail as a result of not being able to read `RepositoryData`